File: nhtsa.py
"""
NHTSA recall + complaints wrapper.
Free public API: api.nhtsa.dot.gov
No auth required.
"""

import logging
from dataclasses import dataclass

import requests

logger = logging.getLogger(__name__)

BASE = "https://api.nhtsa.dot.gov"


@dataclass
class RecallSummary:
    vin: str
    recall_count: int
    open_recalls: list[dict]    # recalls not yet remedied
    complaint_count: int
    risk_flag: bool             # True if open safety-critical recall exists
    summary: str


def check_vin(vin: str) -> RecallSummary:
    """Full risk check for a VIN: recalls + complaint count."""
    recalls = _get_recalls(vin)
    complaints = _get_complaint_count(vin)

    open_recalls = [r for r in recalls if not r.get("RemedyDescription")]
    safety_critical = any(
        any(kw in (r.get("Component", "") + r.get("Summary", "")).upper()
            for kw in ("FUEL", "BRAKE", "STEER", "FIRE", "CRASH", "AIRBAG", "ROLLOVER"))
        for r in open_recalls
    )

    summary = _format_recall_summary(vin, recalls, open_recalls, complaints, safety_critical)

    return RecallSummary(
        vin=vin,
        recall_count=len(recalls),
        open_recalls=open_recalls,
        complaint_count=complaints,
        risk_flag=safety_critical,
        summary=summary,
    )


def _get_recalls(vin: str) -> list[dict]:
    try:
        resp = requests.get(
            f"{BASE}/vehicles/vin/{vin}/recalls",
            timeout=10,
        )
        if resp.status_code == 200:
            return resp.json().get("results", [])
    except Exception as e:
        logger.debug(f"NHTSA recall fetch failed for {vin}: {e}")
    return []


def _get_complaint_count(vin: str) -> int:
    """
    NHTSA complaints are by year/make/model, not VIN.
    Decode VIN first to get year/make/model, then get complaint count.
    """
    decoded = _decode_vin(vin)
    if not decoded:
        return 0

    year = decoded.get("ModelYear", "")
    make = decoded.get("Make", "").upper()
    model = decoded.get("Model", "").upper()

    if not all([year, make, model]):
        return 0

    try:
        resp = requests.get(
            f"{BASE}/complaints/complaintsByVehicle?make={make}&model={model}&modelYear={year}",
            timeout=10,
        )
        if resp.status_code == 200:
            data = resp.json()
            return data.get("count", 0)
    except Exception as e:
        logger.debug(f"NHTSA complaint fetch failed: {e}")
    return 0


def _decode_vin(vin: str) -> dict:
    try:
        resp = requests.get(
            f"https://vpic.nhtsa.dot.gov/api/vehicles/DecodeVin/{vin}?format=json",
            timeout=10,
        )
        if resp.status_code == 200:
            results = resp.json().get("Results", [])
            return {r["Variable"]: r["Value"] for r in results if r.get("Value")}
    except Exception as e:
        logger.debug(f"VIN decode failed: {e}")
    return {}


def _format_recall_summary(
    vin: str,
    recalls: list[dict],
    open_recalls: list[dict],
    complaint_count: int,
    risk_flag: bool,
) -> str:
    flag = "⚠️  RISK FLAG — open safety recall" if risk_flag else "✓  No open safety recalls"

    lines = [
        f"  VIN Check: {vin}",
        f"  {flag}",
        f"  Recalls:    {len(recalls)} total  •  {len(open_recalls)} open",
        f"  Complaints: {complaint_count} on record (NHTSA)",
    ]

    if open_recalls:
        lines.append(f"  Open recall details:")
        for r in open_recalls[:3]:  # show max 3
            component = r.get("Component", "Unknown component")
            summary = r.get("Summary", "")[:80]
            lines.append(f"    • {component}: {summary}...")

    return "\n".join(lines)

File: src/nhtsa.ts
export interface RecallSummary {
  vin: string;
  recallCount: number;
  openRecalls: Record<string, unknown>[];
  complaintCount: number;
  riskFlag: boolean;
  summary: string;
}

const BASE = "https://api.nhtsa.dot.gov";
const VPIC = "https://vpic.nhtsa.dot.gov/api";

async function fetchJson<T>(url: string): Promise<T | null> {
  try {
    const { default: fetch } = await import("node-fetch");
    const res = await fetch(url, { signal: AbortSignal.timeout(10000) });
    if (res.ok) return (await res.json()) as T;
  } catch {}
  return null;
}

async function getRecalls(vin: string): Promise<Record<string, unknown>[]> {
  const data = await fetchJson<{ results: Record<string, unknown>[] }>(
    `${BASE}/vehicles/vin/${vin}/recalls`
  );
  return data?.results ?? [];
}

async function decodeVin(vin: string): Promise<Record<string, string>> {
  const data = await fetchJson<{ Results: Array<{ Variable: string; Value: string }> }>(
    `${VPIC}/vehicles/DecodeVin/${vin}?format=json`
  );
  if (!data) return {};
  return Object.fromEntries(
    data.Results.filter((r) => r.Value).map((r) => [r.Variable, r.Value])
  );
}

async function getComplaintCount(vin: string): Promise<number> {
  const decoded = await decodeVin(vin);
  const year = decoded["ModelYear"];
  const make = decoded["Make"]?.toUpperCase();
  const model = decoded["Model"]?.toUpperCase();
  if (!year || !make || !model) return 0;

  const data = await fetchJson<{ count: number }>(
    `${BASE}/complaints/complaintsByVehicle?make=${make}&model=${model}&modelYear=${year}`
  );
  return data?.count ?? 0;
}

function isSafetyCritical(recall: Record<string, unknown>): boolean {
  const text = `${recall["Component"] ?? ""} ${recall["Summary"] ?? ""}`.toUpperCase();
  return ["FUEL", "BRAKE", "STEER", "FIRE", "CRASH", "AIRBAG", "ROLLOVER"].some((kw) =>
    text.includes(kw)
  );
}

function formatSummary(
  vin: string,
  recalls: Record<string, unknown>[],
  openRecalls: Record<string, unknown>[],
  complaintCount: number,
  riskFlag: boolean
): string {
  const flag = riskFlag
    ? "⚠️  RISK FLAG — open safety recall"
    : "✓  No open safety recalls";

  const lines = [
    `  VIN Check: ${vin}`,
    `  ${flag}`,
    `  Recalls:    ${recalls.length} total  •  ${openRecalls.length} open`,
    `  Complaints: ${complaintCount} on record (NHTSA)`,
  ];

  if (openRecalls.length) {
    lines.push("  Open recall details:");
    for (const r of openRecalls.slice(0, 3)) {
      const component = (r["Component"] as string) ?? "Unknown component";
      const summary = ((r["Summary"] as string) ?? "").slice(0, 80);
      lines.push(`    • ${component}: ${summary}...`);
    }
  }

  return lines.join("\n");
}

export async function checkVin(vin: string): Promise<RecallSummary> {
  const [recalls, complaintCount] = await Promise.all([
    getRecalls(vin),
    getComplaintCount(vin),
  ]);

  const openRecalls = recalls.filter((r) => !r["RemedyDescription"]);
  const riskFlag = openRecalls.some(isSafetyCritical);
  const summary = formatSummary(vin, recalls, openRecalls, complaintCount, riskFlag);

  return { vin, recallCount: recalls.length, openRecalls, complaintCount, riskFlag, summary };
}
